route-rs-runtime/src/link/primitive/pcap.rs

```diff
@@ -0,0 +1,377 @@
+pub struct ToPcap {
+    pcap_file: Option<String>,
+    in_stream: Option<PacketStream<Vec<u8>>>,
+}
+impl Default for ToPcap {
+    fn default() -> Self {
+        ToPcap {
+            pcap_file: None,
+            in_stream: None,
+        }
+    }
+}
+impl ToPcap {
+    pub fn new() -> Self {
+        ToPcap::default()
+    }
+    pub fn pcap_file(self, pcap_file: String) -> Self {
+        ToPcap {
+            pcap_file: Some(pcap_file),
+            in_stream: self.in_stream,
+        }
+    }
+}
+impl LinkBuilder<Vec<u8>, ()> for ToPcap {
+    fn ingressor(self, in_stream: PacketStream<Vec<u8>>) -> Self {
+        if self.in_stream.is_some() {
+            panic!("ToPcap can only take 1 input stream")
+        }
+
+        ToPcap {
+            pcap_file: self.pcap_file,
+            in_stream: Some(in_stream),
+        }
+    }
+    fn ingressors(self, mut ingress_streams: Vec<PacketStream<Vec<u8>>>) -> Self {
+        assert_eq!(
+            ingress_streams.len(),
+            1,
+            "ToPcap can only take 1 ingress stream"
+        );
+
+        if self.in_stream.is_some() {
+            panic!("ToPcap can only take 1 input stream")
+        }
+
+        ToPcap {
+            pcap_file: self.pcap_file,
+            in_stream: Some(ingress_streams.remove(0)),
+        }
+    }
+    fn build_link(self) -> Link<()> {
+        if let Some(pcap_file) = self.pcap_file {
+            if let Some(stream) = self.in_stream {
+                let pcap_proc = ToPcapRunner::new(pcap_file, stream);
+                (vec![Box::new(pcap_proc)], vec![])
+            } else {
+                panic!("Cannot build ToPcap! Missing in_stream")
+            }
+        } else {
+            panic!("Cannot build ToPcap! Missing source file")
+        }
+    }
+}
+
+pub struct ToPcapRunner {
+    #[allow(dead_code)]
+    pcap_file: String,
+    start: Instant,
+    stream: PacketStream<Vec<u8>>,
+    packet_dump: pcap::Savefile,
+}
+impl Unpin for ToPcapRunner {}
+impl ToPcapRunner {
+    pub fn new(pcap_file: String, stream: PacketStream<Vec<u8>>) -> Self {
+        let start = Instant::now();
+        let cap = Capture::dead(Linktype(1)).unwrap();
+        let packet_dump = cap.savefile(pcap_file.clone()).unwrap();
+        ToPcapRunner {
+            pcap_file,
+            start,
+            stream,
+            packet_dump,
+        }
+    }
+}
+impl Future for ToPcapRunner {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let collector = Pin::into_inner(self);
+        loop {
+            match ready!(Pin::new(&mut collector.stream).poll_next(cx)) {
+                Some(packet) => {
+                    let ts = collector.start.elapsed();
+                    let tv_sec = ts.as_secs();
+                    let tv_usec = (ts - Duration::from_secs(tv_sec)).as_micros();
+                    let packet_hdr = PacketHeader {
+                        ts: libc::timeval {
```

Comment:
LOL it is OS-specific.  When I `cargo build --target x86_64-unknown-linux-gnu` it works, doesn't on MacOS.  Could you add a cfg flag for MacOS to be i32 instead of i64? The runtime still supports MacOS somewhat, it should at least build on it.

Reply:
BTW to add targets, you just `rustup target add <triple>`.  List of supported triples are here https://forge.rust-lang.org/release/platform-support.html

Reply:
Graphgen doesn't cross-compile for me right away, I think it needs a C linker for some reason, it threw a wall of text at me in classic linker failure fashion.  But I can cross-compile the runtime for windows, macOS, and Linux. 

Here is a patch that sets the conditional compilation flags appropriately

Reply:
[0001-Runtime-PCAP-Fix-compilation-for-MacOS.patch.txt](https://github.com/route-rs/route-rs/files/4291531/0001-Runtime-PCAP-Fix-compilation-for-MacOS.patch.txt)

Reply:
Applied the patch and made a comment in the git log